golang/cosmos/app/app.go

```diff
@@ -845,26 +844,19 @@ func isFirstTimeUpgradeOfThisVersion(app *GaiaApp, ctx sdk.Context) bool {
 	return true
 }
 
-// upgrade14Handler performs standard upgrade actions plus custom actions for upgrade-14.
-func upgrade14Handler(app *GaiaApp, targetUpgrade string) func(sdk.Context, upgradetypes.Plan, module.VersionMap) (module.VersionMap, error) {
+// upgrade15Handler performs standard upgrade actions plus custom actions for upgrade-15.
+func upgrade15Handler(app *GaiaApp, targetUpgrade string) func(sdk.Context, upgradetypes.Plan, module.VersionMap) (module.VersionMap, error) {
 	return func(ctx sdk.Context, plan upgradetypes.Plan, fromVm module.VersionMap) (module.VersionMap, error) {
 		app.CheckControllerInited(false)
 
 		CoreProposalSteps := []vm.CoreProposalStep{}
 
 		// These CoreProposalSteps are not idempotent and should only be executed
-		// as part of the first upgrade-14 on any given chain.
+		// as part of the first upgrade using this handler on any given chain.
 		if isFirstTimeUpgradeOfThisVersion(app, ctx) {
 			// Each CoreProposalStep runs sequentially, and can be constructed from
 			// one or more modules executing in parallel within the step.
-			CoreProposalSteps = []vm.CoreProposalStep{
-				// First, upgrade wallet factory
-				vm.CoreProposalStepForModules("@agoric/vats/scripts/build-wallet-factory2-upgrade.js"),
-				// Then, upgrade Zoe and ZCF
-				vm.CoreProposalStepForModules("@agoric/vats/scripts/replace-zoe.js"),
-				// Next revive KREAd characters
-				vm.CoreProposalStepForModules("@agoric/vats/scripts/revive-kread.js"),
-			}
+			CoreProposalSteps = []vm.CoreProposalStep{}
 		}
 
```

Comment:
Commenting on lines 887-881 in the new version (silly github won't allow the comment to be added there). This seems like a leftover from upgrade-14 and should be removed, but you'll need to check with whoever added it. In general, we should probably maintain a core function of what should be in every upgrade handler, then the upgrade-N handler calls that and has its own functionality.

```
		m := swingsetkeeper.NewMigrator(app.SwingSetKeeper)
		err = m.MigrateParams(ctx)
		if err != nil {
			return mvm, err
		}
```

I'm also not a fan of the "is first time" logic as it gets us into the dangerous game of not testing what will actually happen on mainnet, but I'll discuss that outside of this PR.

Reply:
> Commenting on lines 877-881 in the new version (silly github won't allow the comment to be added there). This seems like a leftover from upgrade-14 and should be removed, but you'll need to check with whoever added it.
> …
> ```
> 		m := swingsetkeeper.NewMigrator(app.SwingSetKeeper)
> 		err = m.MigrateParams(ctx)
> 		if err != nil {
> 			return mvm, err
> 		}
> ```

That block was [introduced in upgrade-13](https://github.com/Agoric/agoric-sdk/pull/8623) and preserved in upgrade-14, and would be relevant for any parameter changes. I'm inclined to keep it.

> In general, we should probably maintain a core function of what should be in every upgrade handler, then the upgrade-N handler calls that and has its own functionality.

:100:

> I'm also not a fan of the "is first time" logic as it gets us into the dangerous game of not testing what will actually happen on mainnet, but I'll discuss that outside of this PR.

Agreed, but it was necessary for upgrade-14-rc1 and I want to keep this block in case it comes up again.

Reply:
> > I'm also not a fan of the "is first time" logic as it gets us into the dangerous game of not testing what will actually happen on mainnet, but I'll discuss that outside of this PR.
> 
> Agreed, but it was necessary for upgrade-14-rc1 and I want to keep this block in case it comes up again.

There is an issue to remove the need for this hack: https://github.com/Agoric/agoric-sdk/issues/9101

Reply:
Filed #9255 to eventually remove the need for this.